### PR TITLE
[minio] Add types of `composeObject`, `selectObjectContent`, fix `fPutObject` paramter to optional.

### PR DESCRIPTION
**Pull request by seohyun0120**
Hi, I added types of `composeObject()`, `seelctObjectContent()` and fixed 'metaData' paramter of `fPutObject()` to optional. 

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [composeObject](https://docs.min.io/docs/javascript-client-api-reference.html#composeObject), [selectObjectContent](https://docs.min.io/docs/javascript-client-api-reference.html#selectObjectContent), [fPutObject](https://github.com/minio/minio-js/blob/master/examples/put-object.js)
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by OutdatedVersion — CHANGES_REQUESTED** [review]

File: types/minio/index.d.ts [hunk 1]
```diff
@@ -176,6 +176,48 @@ export interface LegalHoldOptions {
...
+    VersionID?: string;
+    MatchETag?: string;
+    NoMatchETag?: string;
+    MatchModifiedSince?: string;
+    MatchUnmodifiedSince?: string;
+    MatchRange?: boolean;
+    Start?: number;
+    End?: number;
+    Encryption?: EmptyObject;
+}
```
[the remark this answers is not in the document]
Comment (seohyun0120) [reply types/minio/index.d.ts]: Thanks, my mistake. I added `CopySourceOptions`, `CopyDestinationOptions`. 
Also I have a question. I typed `CopyDestinationOptions.getHeaders()` return type as `{ [key: string]: any; }` but I'm not sure if this is right. Any better ideas?

File: types/minio/index.d.ts [hunk 2]
```diff
@@ -176,6 +176,48 @@ export interface LegalHoldOptions {
     status: LegalHoldStatus;
 }
 
+export interface SelectOptions {
+    expression: string;
+    expressionType?: string;
+    inputSerialization: EmptyObject;
+    outputSerialization: EmptyObject;
```
[the remark this answers is not in the document]
Comment (seohyun0120) [reply types/minio/index.d.ts]: Thanks for the reference!! 🙇‍♀️  I fixed both serialization types, also fixed `requestProgress`, `scanRange` types following the AWS docs.
- requestProgress https://docs.aws.amazon.com/AmazonS3/latest/API/API_ScanRange.html
- scanRange https://docs.aws.amazon.com/AmazonS3/latest/API/API_RequestProgress.html

I'm not sure how to type `Parquet` property. Any suggestions? 🙏

File: types/minio/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (OutdatedVersion) [reply types/minio/index.d.ts]: > I typed CopyDestinationOptions.getHeaders() return type as { [key: string]: any; } but > I'm not sure if this is right. Any better ideas?

Hhm. I think replacing the `any` with `string` (e.g. `Record<string, string>`) would better illustrate the return value. Maybe throwing in the known headers (like what you did with `CopySourceOptions` already) plus the default index as a fallback is worth the time? e.g. 
```ts
{
    'X-Amz-Tagging-Directive'?: string;
    'X-Amz-Object-Lock-Mode'?: 'GOVERNANCE' | 'COMPLIANCE';
    // ...
    [key: string]: string;
}
```
I don't really mind the known headers being left out though.

File: types/minio/index.d.ts [hunk 3]
```diff
@@ -176,6 +176,55 @@ export interface LegalHoldOptions {
...
+        AllowQuotedRecordDelimiter?: boolean;
+        Comments?: string;
+        FieldDelimiter?: string;
+        FileHeaderInfo?: string;
+        QuoteCharacter?: string;
+        QuoteEscapeCharacter?: string;
+        RecordDelimiter?: string;
+    };
+    JSON?: {
+        [key: string]: any;
```
Comment (OutdatedVersion) [inline types/minio/index.d.ts:191]: Looks like `JSON` takes an optional `Type` property and nothing else: https://docs.aws.amazon.com/AmazonS3/latest/API/API_JSONInput.html

File: types/minio/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (OutdatedVersion) [reply types/minio/index.d.ts]: > I'm not sure how to type Parquet property. Any suggestions? 🙏

I think `Record<string, never>` (or the `EmptyObject` here) would be fine. It was consistently empty looking around some other docs/SDKs.

File: types/minio/index.d.ts [hunk 4]
```diff
@@ -176,6 +176,55 @@ export interface LegalHoldOptions {
     status: LegalHoldStatus;
 }
 
+export interface InputSerialization {
+    CompressionType?: string;
+    CSV?: {
+        AllowQuotedRecordDelimiter?: boolean;
+        Comments?: string;
+        FieldDelimiter?: string;
+        FileHeaderInfo?: string;
```
Comment (OutdatedVersion) [inline types/minio/index.d.ts:185]: nit: 
```suggestion
        FileHeaderInfo?: 'NONE' | 'IGNORE' | 'USE';
```

File: types/minio/index.d.ts [hunk 5]
```diff
@@ -429,6 +484,57 @@ export class CloudFunctionConfig extends TargetConfig {
...
+    constructor(
+        Bucket: string,
+        Object: string,
+        Encryption?: {
+            type: string;
+            SSEAlgorithm?: string;
+            KMSMasterKeyID?: string;
+        },
+        UserMetadata?: EmptyObject,
+        UserTags?: EmptyObject | string,
```
Comment (OutdatedVersion) [inline types/minio/index.d.ts:528]: Here the `EmptyObject`s mean `{}` literally so `Record<string, unknown>` would fit better

**Review by OutdatedVersion — CHANGES_REQUESTED** [review]
I forgot to mention this on the first review, my apologies, but could you bump the version at line 1 of `index.d.ts` to be the latest `minio` release this is compatible with (probably 7.0.27). And throw your name in the authors if you'd like :)

File: types/minio/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (seohyun0120) [reply types/minio/index.d.ts]: > I don't really mind the known headers being left out though.

I agree. I changed both `getHeaders()` return type to `Record<string, string>`.

File: types/minio/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (seohyun0120) [reply types/minio/index.d.ts]: I changed type to `EmptyObject` which is same as `Record<string, never>`.
+ Wow, GitHub Code Search. That looks good.

File: types/minio/index.d.ts (hunk 3 above)
Comment (seohyun0120) [reply types/minio/index.d.ts]: Fixed

File: types/minio/index.d.ts (hunk 4 above)
Comment (seohyun0120) [reply types/minio/index.d.ts]: Fixed!

File: types/minio/index.d.ts (hunk 5 above)
Comment (seohyun0120) [reply types/minio/index.d.ts]: Fixed!

**Comment by seohyun0120** [comment]
> I forgot to mention this on the first review, my apologies, but could you bump the version at line 1 of index.d.ts to be the latest minio release this is compatible with (probably 7.0.27). And throw your name in the authors if you'd like :)

According to [dtslint document](https://github.com/Microsoft/dtslint/blob/master/docs/dt-header.md), patch version is not allowed so cannot change the version. I add my name :)

**Review by OutdatedVersion — CHANGES_REQUESTED** [review]

File: types/minio/index.d.ts [hunk 6]
```diff
@@ -429,6 +483,49 @@ export class CloudFunctionConfig extends TargetConfig {
...
+        MatchUnmodifiedSince?: string,
+        MatchRange?: boolean,
+        Start?: number,
+        End?: number,
+        Encryption?: {
+            type: string;
+            SSEAlgorithm?: string;
+            KMSMasterKeyID?: string;
+        },
+    );
```
[the remark this answers is not in the document]
Comment (seohyun0120) [reply types/minio/index.d.ts]: Oh... That's my mistake. I fixed them.

**Review by OutdatedVersion — APPROVED** [review]

**Comment by seohyun0120** [comment]
Ready to merge

**Pull request merged**